mne/coreg.py

```diff
@@ -1223,3 +1225,528 @@ def _scale_xfm(subject_to, xfm_fname, mri_name, subject_from, scale,
+        mri_pts = list()
+        weights = list()
+        if self.has_hsp_data and self.hsp_weight > 0:  # should be true
+            head_pts.append(self.hsp.points)
+            mri_pts.append(self.processed_high_res_mri_points[
+                self.nearest_transformed_high_res_mri_idx_hsp])
+            weights.append(np.full(len(head_pts[-1]), self.hsp_weight))
+        for key in ('lpa', 'nasion', 'rpa'):
+            if getattr(self, 'has_%s_data' % key):
+                head_pts.append(getattr(self.hsp, key))
+                if self.icp_fid_match == 'matched':
+                    mri_pts.append(getattr(self.mri, key))
+                else:
+                    assert self.icp_fid_match == 'nearest'
+                    mri_pts.append(self.processed_high_res_mri_points[
+                        getattr(self, 'nearest_transformed_high_res_mri_idx_%s'
+                                % (key,))])
+                weights.append(np.full(len(mri_pts[-1]),
+                                       getattr(self, '%s_weight' % key)))
+        if self.has_eeg_data and self.eeg_weight > 0:
+            head_pts.append(self.hsp.eeg_points)
+            mri_pts.append(self.processed_high_res_mri_points[
+                self.nearest_transformed_high_res_mri_idx_eeg])
+            weights.append(np.full(len(mri_pts[-1]), self.eeg_weight))
+        if self.has_hpi_data and self.hpi_weight > 0:
+            head_pts.append(self.hsp.hpi_points)
+            mri_pts.append(self.processed_high_res_mri_points[
+                self.nearest_transformed_high_res_mri_idx_hpi])
+            weights.append(np.full(len(mri_pts[-1]), self.hpi_weight))
+        head_pts = np.concatenate(head_pts)
+        mri_pts = np.concatenate(mri_pts)
+        weights = np.concatenate(weights)
+        if self.n_scale_param == 0:
+            mri_pts *= self.parameters[6:9]  # not done in fit_matched_points
+        return head_pts, mri_pts, weights
+
+    def fit_icp(self, iterations=20, lpa_weight=1., nasion_weight=10.,
+                rpa_weight=1.):
+        """Find MRI scaling, translation, and rotation to match HSP."""
+        self.lpa_weight = lpa_weight
+        self.nasion_weight = nasion_weight
+        self.rpa_weight = rpa_weight
+
+        # Initial guess (current state)
+        est = self.parameters[:[6, 7, None, 9][self.n_scale_param]]
+
+        # Do the fits, assigning and evaluating at each step
+        for _ in range(iterations):
+            head_pts, mri_pts, weights = self._setup_icp()
+            est = fit_matched_points(mri_pts, head_pts,
+                                     scale=self.n_scale_param,
+                                     x0=est, out='params', weights=weights)
+            if self.n_scale_param == 0:
+                self.parameters[:6] = est
+            elif self.n_scale_param == 1:
+                self.parameters[:] = list(est) + [est[-1]] * 2
+            else:
+                self.parameters[:] = est
```

Comment:
is it documentated what self.parameters can contain depending on self.n_scale_param?

why not using out='trans' and just work with the trans?

@larsoner any opinion?

Reply:
> is it documentated what self.parameters can contain depending on self.n_scale_param?

No but it can be figured out from the code and documented in this PR. IIRC it's rotation + quats + scale factors (so 6, 7, or 9 params)

> why not using out='trans' and just work with the trans?

Loss of precision and internal computations / optimization is done on quaternions so better just to quat_to_rot on the fly when necessary (it's very cheap in the grand scheme of ICP or viz)